### PR TITLE
Enhance #386 and add #247

### PR DESCRIPTION
**Pull request by geauxvirtual**
Enhances #386 - Global Config Spec for Snapd
Feature #247 - Exposes WorkManagerQueueSize and WorkManagerPoolSize settings via Command line flags and configuration files.

Summary of changes:
- Removes pkg/globalconfig from handling reading configuration files. This package did not implement #386 per the discussed spec and also included specific snapd information which did not follow the pattern for packages added to pkg directory.
- Adds pkg/cfgfile for reading YAML and JSON configuration files. This package could be extended to support other configuration file types.
- Snapd supports configuration files in YAML or JSON format. There are five sections to a snapd configuration: Top-level (configuration flags strictly for snapd), Control, Scheduler, RestAPI, and Tribe. Configuration settings for each module (Control, Scheduler, RestAPI, and Tribe) are now located under each section, as shown in the included example configuration files. Plugin configuration is under the Control section. 
- Each module defines a default configuration value for each configuration setting the module exposes. Default values have been removed from the defined flags. A `GetDefaultConfig()` method exists for each module to return the default configuration for that module.
- Snapd now defines the default configuration file location as `/etc/snap/snapd.conf`. If this file exists, snapd will attempt to read it and load the configuration defined in this file. If a configuration file is passed with the `--config` flag, this configuration takes precedence over the default configuration file if present. 
- Values defined in a configuration file take precedence over default values.
- Any values defined by a command flag on start of snapd takes precedence over default values or values defined in a configuration file.

Example default YAML configuration file:
```yaml
---
# log_level for the snap daemon. Supported values are
# 1 - Debug, 2 - Info, 3 - Warning, 4 - Error, 5 - Fatal.
# Default value is 3.
log_level: 1

# log_path sets the path for logs for the snap daemon. By
# default snapd prints all logs to stdout. Any provided
# path will send snapd logs to a file called snapd.log in
# the provided directory.
#log_path: /some/log/dir

# Gomaxprocs sets the number of cores to use on the system
# for snapd to use. Default for gomaxprocs is 1
gomaxprocs: 4

# Control sections for configuration settings for the plugin
# control module of snapd.
control:
  # auto_discover_path sets a directory to auto load plugins on the start
  # of the snap daemon
  #auto_discover_path: /some/directory/with/plugins

  # cache_expiration sets the time interval for the plugin cache to use before
  # expiring collection results from collect plugins. Default value is 500ms
  cache_expiration: 750ms

  # max_running_plugins sets the size of the available plugin pool for each
  # plugin loaded in the system. Default value is 3
  max_running_plugins: 1

  # keyring_paths sets the directory(s) to search for keyring files for signed
  # plugins. This can be a comma separated list of directories
  #keyring_paths: /some/path/with/keyring/files

  # plugin_trust_level sets the plugin trust level for snapd. The default state
  # for plugin trust level is enabled (1). When enabled, only signed plugins that can
  # be verified will be loaded into snapd. Signatures are verifed from
  # keyring files specided in keyring_path. Plugin trust can be disabled (0) which
  # will allow loading of all plugins whether signed or not. The warning state allows
  # for loading of signed and unsigned plugins. Warning messages will be displayed if
  # an unsigned plugin is loaded. Any signed plugins that can not be verified will
  # not be loaded. Valid values are 0 - Off, 1 - Enabled, 2 - Warning
  plugin_trust_level: 0

  # plugins section contains plugin config settings that will be applied for
  # plugins across tasks.
  plugins:
    all:
      password: p@ssw0rd
    collector:
      all:
        user: jane
      pcm:
        all:
          path: /usr/local/pcm/bin
        versions:
          1:
            user: john
            someint: 1234
            somefloat: 3.14
            somebool: true
      psutil:
        all:
          path: /usr/local/bin/psutil
    publisher:
      influxdb:
        all:
          server: xyz.local
          password: $password
    processor:
      movingaverage:
        all:
          user: jane
        versions:
          1:
            user: tiffany
            password: new password

# scheduler configuration settings contains all settings for scheduler
# module
# scheduler:
  # work_manager_queue_size sets the size of the worker queue inside snapd scheduler.
  # Default value is 25.
  #work_manager_queue_size: 10

  # work_manager_pool_size sets the size of the <blank>. Default value is 4.
  #work_manager_pool_size: 2

# rest sections contains all the configuration items for the Rest API server.
restapi:
  # enable controls enabling or disabling the Rest API for snapd. Default value is enabled.
  enable: true

  # https enables HTTPs for the Rest API. If no default certificate and key are provided, then
  # the Rest API will generate a private and public key to use for communication. Default
  # value is false
  https: true

  # rest_auth enables authentication for the Rest API. Default value is false
  #rest_auth: true

  # rest_auth_password sets the password to use for the Rest API. Currently user and password
  # combinations are not supported.
  #rest_auth_password: changeme

  # rest_certificate is the path to the certificate to use for Rest API when HTTPS is also enabled.
  #rest_certificate: /path/to/cert/file

  # rest_key is the path to the private key for the certificate in use by the Rest API
  # when HTTPs is enabled.
  #rest_key: /path/to/private/key

  # port sets the port to start the Rest API server on. Default is 8181
  port: 8282

# tribe section contains all configuration items for the tribe module
tribe:
  # enable controls enabling tribe for the snapd instance. Default value is false.
  enable: false

  # bind_addr sets the IP address for Tribe to bind.
  #bind_addr: 127.0.0.1

  # bind_port sets the port for tribe to listen on. Default value is 6000
  #bind_port: 16000

  # name sets the name to use for this snapd instance in the tribe
  # membership. Default value defaults to local hostname of the system.
  #name: localhost

  # seed sets the snapd instance to use as the seed for tribe communications
  #seed: 1.1.1.1:16000
```

Starting snapd with a default configuration file:
screen shot 2016-03-10 at 1 58 03 pm

Example JSON configuration file passed in:
```json
{
    "log_level": 1,
    "control": {
        "auto_discover_path": "build/plugin",
        "cache_expiration": "1s",
        "max_running_plugins": 3,
        "plugin_trust_level": 0,
    },
    "restapi": {
        "enable": true,
    }
}
```

Starting snapd with a configuration file passed in on the command line via `--config`:
screen shot 2016-03-14 at 9 52 50 am

Starting snapd with default configuration file but passing in flags to override settings:
screen shot 2016-03-10 at 2 19 16 pm

Starting snapd with `--config` flag and overriding settings in that config file with other flags:
screen shot 2016-03-14 at 9 53 51 am

- Added full example configuration files in YAML (with comments on each setting) and JSON.
- Added tests in pkg/cfgfile for parsing YAML and JSON files.
- Added tests for parsing and validating example configuration files located at `examples/configs/snap-config-sample.json` and `examples/configs/snap-config-sample.yaml`
- Added tests for validating default configuration defined in each module.

- Adds #247: WorkManagerQueueSize and WorkManagerPoolSize configuration via command line or configuration file.

Testing done:
- Validated all tests pass when running `make tests`
- Validated snapd will run without a default configuration file defined in `/etc/snap/snapd.conf`
- Validated precedence order of configuration values:
   1. Flags
   2. Configuration file passed on command line
   3. Default configuration file set at /etc/snap/snapd.conf
   4. Default configuration value defined in code

@intelsdi-x/snap-maintainers

### CODE REVIEW

**Comment by ConnorDoyle** [comment]
First of all, this is awesome.  Just had an idea about extra tooling around configurations.  Would it be helpful to have a flag that causes `snapd` to validate and dump the effective configuration (return nonzero on error)?

**Comment by tjmcs** [comment]
I think that would be a great enhancement; were you thinking of a "debug_startup" flag of some sort that would describe the configuration that would be applied based on a given snapd command (after applying defaults, the appropriate configuration flag, if any, and any command line flags that were provided) or were you thinking of a command you could run from the CLI or RESTful API to retrieve the current configuration of a running snapd instance?

**Comment by geauxvirtual** [comment]
What would be considered the role of validate? Just that there is no errors in parsing a YAML or JSON configuration file? Or also validate values passed in as being acceptable? (i.e MaxRunningPlugins is between 1 and 10)

**Comment by ConnorDoyle** [comment]
Moved config validation discussion to issue #764 to avoid hijacking this review.

**Comment by lynxbat** [comment]
This is a wicked PR. Probably one of my favs recently.

That said. Could we maybe get a doc added and linked to the readme on how this config works with examples and a reference for each config item? One of my favorite examples of this is the Apache Mesos one (http://mesos.apache.org/documentation/latest/configuration/)

**Comment by lynxbat** [comment]
Pinging @mjbrender to weigh in on the above ^

**Comment by danielscottt** [comment]
@lynxbat I came here to say, " I think you're supposed to use 'reactions' for this."  I turns out you already had, so :+1:.

**Comment by geauxvirtual** [comment]
I agree, there should be a documentation page detailing the configuration file.

**Comment by mjbrender** [comment]
Absolutely @lynxbat @geauxvirtual @danielscottt. We're to a point where documentation could be expanded out. I'll open two issues - one to write up global config as a md file in docs and another to consider how to serve our expanding amounts of docs in a better way.

**Comment by lynxbat** [comment]
I wouldn't want this merged without the doc explaining it.

**Comment by geauxvirtual** [comment]
I've added a docs/SNAPD_CONFIGURATION.md page to the PR that is available for review now.

**Comment by lynxbat** [comment]
Awesome. I'll let @mjbrender do the doc review since I am the worst at it :)

File: docs/SNAPD_CONFIGURATION.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# snapd Configuration File
+
+snapd supports being configured through a configuration file located at a default location of `/etc/snap/snapd.conf` on Linux systems or by passing a configuration file in through the `--config` command line flag when starting snapd. YAML and JSON are currently supported for configuration file types.
+
+snapd will start and run without a provided configuration file and the default values defined inside the daemon will be used. There is an order of precedence when it come to default values, configuration files, and flags when snapd starts. Any value defined in the default configuration file located at `/etc/snap/snapd.conf` will take precedence over default values. Any value defined in a configuration file passed via the `--config` flag will be used in place of any default configuration file on the system and override default values. Any flags passed in on the command line during the start up of snapd will override any values defined in configuration files and default values.
```
Comment (mjbrender) [inline docs/SNAPD_CONFIGURATION.md]: I found this sentence confusing. Rewording from: 
> snapd will start and run without a provided configuration file and the default values defined inside the daemon will be used.

To: 
> snapd runs without a configuration file provided and will use the default values defined inside the daemon (shown below).

File: docs/SNAPD_CONFIGURATION.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+In order of precedence:
```
Comment (mjbrender) [inline docs/SNAPD_CONFIGURATION.md]: In order of precedence (greatest to least):

File: docs/SNAPD_CONFIGURATION.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  # keyring files specided in keyring_path. Plugin trust can be disabled (0) which
+  # will allow loading of all plugins whether signed or not. The warning state allows
+  # for loading of signed and unsigned plugins. Warning messages will be displayed if
+  # an unsigned plugin is loaded. Any signed plugins that can not be verified will
+  # not be loaded. Valid values are 0 - Off, 1 - Enabled, 2 - Warning
+  plugin_trust_level: 1
+  
+  # plugins section contains plugin config settings that will be applied for
+  # plugins across tasks.
+  plugins:
```
Comment (mjbrender) [inline docs/SNAPD_CONFIGURATION.md]: More a note to self - this a very cool feature that will need to have end-to-end examples. :ok_hand:

File: docs/SNAPD_CONFIGURATION.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+### snapd scheduler configurations
+The scheduler section of the configuration file configures settings for the Scheduler module inside the snap daemon.
+
+```yaml
+scheduler:
+  # work_manager_queue_size sets the size of the worker queue inside snapd scheduler.
+  # Default value is 25.
+  work_manager_queue_size: 25
+
+  # work_manager_pool_size sets the size of the <blank>. Default value is 4.
```
Comment (mjbrender) [inline docs/SNAPD_CONFIGURATION.md]: intentionally left <blank>?

Comment (geauxvirtual) [inline docs/SNAPD_CONFIGURATION.md]: Can you expand on this comment?

File: docs/SNAPD_CONFIGURATION.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+### snapd rest API configurations
+The restapi section of the configuration file configures settings for enabling and running the Rest API as part of the snap daemon. The snapctl command line tool uses the Rest API to manage snapd on a host. 
+
+```yaml
+restapi:
+  # enable controls enabling or disabling the Rest API for snapd. Default value is enabled.
+  enable: true
+
+  # https enables HTTPs for the Rest API. If no default certificate and key are provided, then
```
Comment (mjbrender) [inline docs/SNAPD_CONFIGURATION.md]: Very minor, but I've always seen HTTPS with a capitol 's'

File: docs/SNAPD_CONFIGURATION.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+  port: 8181
+```
+
+### snapd tribe configurations
+The tribe section of the configuration file configures settings for enabling and running tribe as part of the snap daemon.
+```yaml
+tribe:
+  # enable controls enabling tribe for the snapd instance. Default value is false.
+  enable: false
+
```
Comment (mjbrender) [inline docs/SNAPD_CONFIGURATION.md]: s/Tribe/tribe/

File: docs/SNAPD_CONFIGURATION.md [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+  # bind_addr sets the IP address for Tribe to bind.
```
Comment (mjbrender) [inline docs/SNAPD_CONFIGURATION.md]: s/Tribe/tribe/

File: docs/SNAPD_CONFIGURATION.md (hunk 4 above)
Comment (mjbrender) [inline docs/SNAPD_CONFIGURATION.md]: The line `# work_manager_pool_size sets the size of the <blank>. Default value is 4.` I wasn't sure: 

* what `<blank>` was supposed to be
* how much description is worth adding? 

@tjmcs added a little context that does the trick!

Comment (geauxvirtual) [inline docs/SNAPD_CONFIGURATION.md]: Oh, right. I was going to ask @danielscottt for what should replace the <blank>.

Comment (danielscottt) [inline docs/SNAPD_CONFIGURATION.md]: `work_manager_pool_size` is the size of the thread (goroutine) pool on the receiving end of the `work_manager_queue`.

**Comment by jcooklin** [comment]
LGTM :ship: it

**Pull request merged**